# JUICE JUNCTION PostgreSQL QUERIES~

Installing the required Libraries.

In [1]:
!pip install psycopg2-binary sqlalchemy pandas

In [2]:
import pandas as pd
from sqlalchemy import create_engine

# PostgreSQL connection details
HOST = "localhost"      
PORT = "5432"           
DATABASE = "juicejunction"
USER = "postgres"
PASSWORD = "Geniusme2@"

# Create SQLAlchemy Engine

engine = create_engine("postgresql://postgres:Geniusme2%40@localhost:5432/juicejunction")



In [3]:
# Load dataset


df = pd.read_csv("C:/Users/chari/OneDrive/Desktop/MY WORK/Juice-Junction-Buisness/Juice_Junction_buisness_sales.csv")  
df.head()


Order Date     Category                  Item Product_ID  \
0     Monday, 21 August 2023        Chats             Bhel Puri       P006   
1      Sunday, 4 August 2024  Fresh Juice           Mango Juice       P010   
2  Tuesday, 20 February 2024        Chats           Masala Puri       P007   
3    Thursday, 25 April 2024       Shakes       Chocolate Shake       P013   
4   Friday, 10 November 2023    Ice Cream  Strawberry Ice Cream       P020   

   Quantity  Price   Total Sales  Customer_ID Payment_Type  Inventory_Stock  
0         4     75            300       C4766       Wallet               91  
1         4     90            360       C5224         Cash               30  
2         4    137            548       C5858         Card               35  
3         3     88            264       C9902       Wallet               47  
4         5    143            715       C4825       Wallet                8

In [4]:
# Push dataframe to PostgreSQL


table_name = "juicejunction_sales"  

df.to_sql(table_name, engine, if_exists='replace', index=False)  

print("Data uploaded successfully!")


Data uploaded successfully!


# Total Revenue Generated Categorically

In [58]:
query = """
SELECT "Category",SUM(" Total Sales ") AS "Total Sales"
FROM juicejunction_sales
GROUP BY "Category"
ORDER BY "Total Sales" DESC;
"""

result = pd.read_sql(query, engine)
result


Category  Total Sales
0       Snacks     365839.0
1        Chats     357791.0
2  Fresh Juice     352266.0
3    Ice Cream     348950.0
4       Shakes     345447.0

# Top 3 Customers with highest total purchase

In [6]:
query = """
SELECT "Customer_ID",SUM(" Total Sales ") AS "Total Purchase"
FROM juicejunction_sales
GROUP BY "Customer_ID"
ORDER BY "Total Purchase" DESC
LIMIT 3;
"""

result = pd.read_sql(query, engine)
result


Customer_ID  Total Purchase
0       C9432          2185.0
1       C4689          2010.0
2       C6327          1956.0

# Top 3 Items with highest total sales

In [7]:
query = """
SELECT "Item",SUM(" Total Sales ") AS "Total Sales"
FROM juicejunction_sales
GROUP BY "Item"
ORDER BY "Total Sales" DESC
LIMIT 3;
"""

result = pd.read_sql(query, engine)
result

Item  Total Sales
0  French Fries      97604.0
1   Mango Shake      97562.0
2     Pani Puri      96107.0

# Monthly Sales trend for 2023

In [8]:
query = """
SELECT DATE_TRUNC('month',"Order Date"::date) AS "Month", SUM(" Total Sales ") AS "Total Revenue"
FROM juicejunction_sales
WHERE EXTRACT(YEAR FROM "Order Date"::date)  = '2023'
GROUP BY "Month"
ORDER BY "Month";
"""

result = pd.read_sql(query, engine)
result

Month  Total Revenue
0  2022-12-31 18:30:00+00:00        80975.0
1  2023-01-31 18:30:00+00:00        62078.0
2  2023-02-28 18:30:00+00:00        73360.0
3  2023-03-31 18:30:00+00:00        72680.0
4  2023-04-30 18:30:00+00:00        74846.0
5  2023-05-31 18:30:00+00:00        69314.0
6  2023-06-30 18:30:00+00:00        89903.0
7  2023-07-31 18:30:00+00:00        69879.0
8  2023-08-31 18:30:00+00:00        73818.0
9  2023-09-30 18:30:00+00:00        73272.0
10 2023-10-31 18:30:00+00:00        69259.0
11 2023-11-30 18:30:00+00:00        75827.0

# Monthly Sales Trend for 2024

In [9]:
query = """
SELECT DATE_TRUNC('month',"Order Date"::date) AS "Month", SUM(" Total Sales ") AS "Total Revenue"
FROM juicejunction_sales
WHERE EXTRACT(YEAR FROM "Order Date"::date)  = '2024'
GROUP BY "Month"
ORDER BY "Month";
"""

result = pd.read_sql(query, engine)
result

Month  Total Revenue
0  2023-12-31 18:30:00+00:00        65666.0
1  2024-01-31 18:30:00+00:00        67265.0
2  2024-02-29 18:30:00+00:00        81603.0
3  2024-03-31 18:30:00+00:00        78839.0
4  2024-04-30 18:30:00+00:00        76202.0
5  2024-05-31 18:30:00+00:00        69160.0
6  2024-06-30 18:30:00+00:00        74013.0
7  2024-07-31 18:30:00+00:00        75206.0
8  2024-08-31 18:30:00+00:00        70858.0
9  2024-09-30 18:30:00+00:00        82312.0
10 2024-10-31 18:30:00+00:00        61604.0
11 2024-11-30 18:30:00+00:00        82354.0

# Weeday Vs Weekend Turnover

In [14]:
query = """
SELECT CASE
       WHEN EXTRACT(ISODOW FROM "Order Date"::date) IN (6,7) THEN 'Weekend'
       ELSE 'Weekday'
       END AS "Order Day", 
       SUM(" Total Sales ") AS "Total Revenue"
FROM juicejunction_sales
GROUP BY "Order Day"
ORDER BY "Total Revenue" DESC;
"""

result = pd.read_sql(query, engine)
result

Order Day  Total Revenue
0   Weekday      1258696.0
1   Weekend       511597.0

# Customer Segmentation

## My active customers are those who have done at least one purchase within the last 6 months from the last order date. Customers beyond the one year line are churned.

#LAST ORDER DATE

In [55]:
query = """
(SELECT MAX("Order Date") FROM juicejunction_sales);
"""

result = pd.read_sql(query, engine)
result

max
0  Wednesday, 9 October 2024

In [57]:
query = """

WITH "Churned Customers" AS 
(  WITH "Last Purchase" AS
    (SELECT DISTINCT("Customer_ID"),
           "Order Date",
           MIN("Order Date") AS "First Purchase Date",
           ROW_NUMBER() OVER (PARTITION BY "Customer_ID" ORDER BY "Order Date" DESC) AS rn,
           SUM(" Total Sales ") AS "Total Revenue"
    FROM juicejunction_sales
    GROUP BY "Order Date","Customer_ID"
    ORDER BY "Total Revenue" DESC)

SELECT "Customer_ID", "Order Date" AS "Last Purchase Date",
        CASE 
        WHEN "Order Date"::date < (SELECT MAX("Order Date") FROM juicejunction_sales)::date - INTERVAL '6 months' THEN 'Churned'
        ELSE 'Active'
        END AS "Customer Status"
FROM "Last Purchase"
WHERE "rn" = 1)


SELECT "Customer Status", COUNT("Customer_ID") AS "Num of Customers", SUM(COUNT("Customer_ID")) OVER () AS "Total Customers",
ROUND((COUNT("Customer_ID") / SUM(COUNT("Customer_ID")) OVER ()),2) AS "Percentage"
FROM "Churned Customers"
GROUP BY "Customer Status"
    
"""

result = pd.read_sql(query, engine)
result

Customer Status  Num of Customers  Total Customers  Percentage
0          Active              1802           4858.0        0.37
1         Churned              3056           4858.0        0.63

# Overall Analysis Insights:

1) Top 3 Categories generating high revenue - Snacks, Chats & Fresh Juice.
2) Top 3 Items - French Fries, Mango shake & Pani puri.
3) Weekday Revenue is more than Weekends.
4) More Customer churn rate.

Business Interpretation:

1. High Churn Rate (63%)

This is a warning sign — most of your customers aren’t staying loyal. That implies:

Customer retention is weak.

Acquisition costs may be wasted because users aren’t sticking around.

There could be issues with product satisfaction, pricing, support, or user experience.

2. Active Customers (37%)

37% of active customers represent your core base.
These are most valuable users — the ones to:

Nurture with loyalty programs or exclusive offers.

Upsell or cross-sell products to.